pkg/resiliency/resiliency.go

```diff
@@ -1,132 +1,508 @@
-// Copyright (c) Microsoft Corporation and Dapr Contributors.
-// Licensed under the MIT License.
-// ------------------------------------------------------------
+/*
+Copyright 2021 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
 
 package resiliency
 
 import (
 	"context"
+	"encoding/json"
+	"fmt"
 	"os"
 	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
 	"time"
 
 	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	lru "github.com/hashicorp/golang-lru"
 	"gopkg.in/yaml.v2"
 
 	resiliency_v1alpha "github.com/dapr/dapr/pkg/apis/resiliency/v1alpha1"
 	operatorv1pb "github.com/dapr/dapr/pkg/proto/operator/v1"
 
+	"github.com/dapr/dapr/pkg/resiliency/breaker"
+	"github.com/dapr/kit/config"
 	"github.com/dapr/kit/logger"
+	"github.com/dapr/kit/retry"
 )
 
 const (
 	operatorRetryCount   = 100
 	operatorTimePerRetry = time.Second * 5
 )
 
-// Wrapper to let us define some helper functions without duplicating the domain definition.
-type Resiliency struct {
-	Resiliency resiliency_v1alpha.Resiliency
-}
+// ActorCircuitBreakerScope indicates the scope of the circuit breaker for an actor.
+type ActorCircuitBreakerScope int
 
-func LoadDefaultResiliency() *Resiliency {
-	return &Resiliency{resiliency_v1alpha.Resiliency{}}
-}
+const (
+	// ActorCircuitBreakerScopeType indicates the type scope (less granular).
+	ActorCircuitBreakerScopeType ActorCircuitBreakerScope = iota
+	// ActorCircuitBreakerScopeID indicates the type+id scope (more granular).
+	ActorCircuitBreakerScopeID
+	// ActorCircuitBreakerScopeBoth indicates both type and type+id are used for scope.
+	ActorCircuitBreakerScopeBoth // Usage is TODO
+)
+
+type (
+	Provider interface {
+		RoutePolicy(ctx context.Context, name string) Runner
+		EndpointPolicy(ctx context.Context, service string, endpoint string) Runner
+		RemoveEndpoint(name string, endpoint string)
+		ActorPolicy(ctx context.Context, actorType string, id string) Runner
+		ComponentPolicy(ctx context.Context, name string) Runner
+	}
+
+	// Resiliency encapsulates configuration for timeouts, retries, and circuit breakers.
+	// It maps services, actors, components, and routes to each of these configurations.
+	// Lastly, it maintains circuit breaker state across invocations.
+	Resiliency struct {
+		timeouts        map[string]time.Duration
+		retries         map[string]*retry.Config
+		circuitBreakers map[string]*breaker.CircuitBreaker
+
+		actorCBCaches map[string]*lru.Cache
```

Comment:
Yes. Since there is an unbounded amount of actors, the cache size is a mechanism to evict older circuit breaker state. We might also apply this approach to endpoints instead of explicitly removing them from service via `RemoveEndpoint`.

Reply:
I think it'd be good to have both under an LRU just to reduce the number of concepts that we're tracking here. Thoughts?

Reply:
Both as in also have Endpoints under an LRU? I can get behind that.